Mas d31 nhskv16sstpcl

martinsumner opened the pull request:
A refactoring of the keyfolder function in the leveled_penciller.  Looking at the eprof profile, this dominates the use of CPU during queries.

The refactoring is:
- Change keyfolder function to treat the in-memory portion of the iterator like the persisted part.  This doesn't change performance, but greatly reduces the line count of code necessary for this function.
- Change keyfolder/find_nextkey to use a map not a [{Level, KeyList}] - replacing keyfind/keyreplace with map get/update.
- Combines comparison functions used into one leveled_codec:maybe_accumulate function

martinsumner committed:
Update leveled_penciller.erl

martinsumner committed:
Refactor query manifest to use map

Avoid keyfind/keyreplace overheads

martinsumner committed:
Refactor keyfolder for speed

martinsumner committed:
Strip out outer loop

This is slightly slower, to compare IMM to SSTM together in the same loop - but large reduction in code lines.

martinsumner committed:
Update with test

There was no test of queries with both modified ranges and expiring keys.

martinsumner commented:
Query PROFILE - develop-3.1:

```
leveled_runner:add_terms/2                           20005743     1.44    726349  [      0.04]

leveled_codec:from_ledgerkey/1                       20005743     1.47    743308  [      0.04]

leveled_penciller:find_nextkey/5                     20006606     1.47    743625  [      0.04]

lists:keyreplace/4                                   20025643     1.48    746753  [      0.04]

lists:keyreplace3/4                                  20943786     1.51    764243  [      0.04]

leveled_codec:strip_to_statusonly/1                  20005743     1.59    805196  [      0.04]

leveled_codec:strip_to_indexdetails/1                20005743     1.76    888410  [      0.04]

leveled_codec:is_active/3                            20005743     2.87   1453401  [      0.07]

leveled_runner:'-accumulate_index/3-fun-0-'/5        20005743     4.31   2181537  [      0.11]

leveled_penciller:keyfolder/4                        20006606     4.67   2362360  [      0.12]

leveled_penciller:maybe_accumulate/5                 20005743     5.93   3001157  [      0.15]

erlang:binary_to_term/1                                392447    14.65   7412605  [     18.89]

lists:keyfind/3                                     220096019    17.43   8823828  [      0.04]

leveled_penciller:find_nextkey/8                    220092814    33.58  16995070  [      0.08]

--------------------------------------------------  ---------  -------  --------  [----------]

Total:                                              696736775  100.00%  50614661  [      0.07]
```


Query PROFILE - with mas-d31-nhskv16sst:

```
lists:keyreplace/4                                   20036099     1.40    717059  [      0.04]

leveled_runner:add_terms/2                           20015189     1.40    719653  [      0.04]

leveled_codec:strip_to_statusonly/1                  20015189     1.43    733119  [      0.04]

leveled_codec:from_ledgerkey/1                       20015189     1.44    739140  [      0.04]

leveled_penciller:find_nextkey/5                     20016129     1.47    752953  [      0.04]

leveled_codec:strip_to_indexdetails/1                20015189     1.62    828220  [      0.04]

leveled_codec:is_active/3                            20015189     2.87   1471794  [      0.07]

lists:keyreplace3/4                                  39064435     3.98   2038030  [      0.05]

leveled_runner:'-accumulate_index/3-fun-0-'/5        20015189     4.24   2172372  [      0.11]

leveled_penciller:keyfolder/4                        20016129     4.68   2400639  [      0.12]

leveled_penciller:maybe_accumulate/5                 20015189     5.82   2984698  [      0.15]

erlang:binary_to_term/1                                395254    14.25   7304877  [     18.48]

lists:keyfind/3                                     220202090    17.31   8869397  [      0.04]

leveled_penciller:find_nextkey/8                    220198446    33.10  16966702  [      0.08]

--------------------------------------------------  ---------  -------  --------  [----------]

Total:                                              711934554  100.00%  51252189  [      0.07]
```

Query PROFILE - with mas-d31-nhskv16sst + this PR:

```
leveled_penciller:'-find_nextkey/8-fun-0-'/1         20009934     2.77    708614  [      0.04]

leveled_runner:add_terms/2                           20009934     2.80    715078  [      0.04]

leveled_codec:from_ledgerkey/1                       20009934     3.08    787532  [      0.04]

leveled_penciller:compare_nextkey/2                  25294142     4.63   1182098  [      0.05]

maps:update_with/3                                   20010856     5.70   1454430  [      0.07]

leveled_penciller:keyfolder/5                        20010856     8.45   2157117  [      0.11]

leveled_codec:maybe_accumulate/5                     40019868     8.56   2186778  [      0.05]

leveled_runner:'-accumulate_index/3-fun-0-'/5        20009934     8.80   2246342  [      0.11]

leveled_penciller:find_nextkey/8                     84177249    18.54   4734316  [      0.06]

erlang:binary_to_term/1                                394735    28.96   7395211  [     18.73]

--------------------------------------------------  ---------  -------  --------  [----------]

Total:                                              297332775  100.00%  25537140  [      0.09]
```

martinsumner commented:
Testing on 8M and 16M object stores with 32M/64M index entries, with a variety of index queries (which count as their accumulator).

This PR gives a 22% reduction in overall latency for 2i queries on top of the benefits of https://github.com/martinsumner/leveled/pull/425.  No detriment seen with any other function.

Combining these PRs, and with an OPT 22 to OTP 26 upgrade - gives an overall reduction in query time of 54%.

martinsumner committed:
Fix test issue

martinsumner committed:
Simplify maybe_accumulate

Explicitly handle the v1 object format, without requiring a second loop.

Also keyfolder/find_next_key sort their tuples at the start and don't re-arrange between them.  Mutable inputs are more clearly separated from constant ones.

martinsumner committed:
Give accumulate_index knowledge of key format

This avoids multiple functions to be required.  The profile indicates this improves performance, but that actual load testing shows no significant difference

martinsumner committed:
Allow for keyfolder to request mutliple keys

This reduces the loops around the keyfolder.

Profiling tool indicates significant saving due to significantly reduced function calls in keyfolder and accumulators.  However, in more real world scenarios, performance seems constant.

Query profile (2M keys):

```
erlang:'--'/2                                         3121553     0.46    179142  [      0.06]

leveled_penciller:keyfolder/6                         1252079     0.70    270316  [      0.22]

erlang:'++'/2                                         1050774     1.54    595734  [      0.57]

perf_SUITE:'-random_queries/5-fun-0-'/3              40020670     3.62   1405230  [      0.04]

leveled_penciller:'-find_nextkeys/6-fun-1-'/1        38770487     3.74   1449232  [      0.04]

leveled_codec:'-accumulate_index/2-fun-1-'/4         40020670     4.07   1577706  [      0.04]

leveled_penciller:compare_nextkey/2                  55594956     7.51   2911541  [      0.05]

maps:update_with/3                                   40021618     7.94   3079187  [      0.08]

leveled_codec:maybe_accumulate/5                     41271801     7.99   3097875  [      0.08]

leveled_penciller:find_nextkeys/6                    98779474    20.73   8040614  [      0.08]

erlang:binary_to_term/1                                786535    36.49  14157118  [     18.00]

--------------------------------------------------  ---------  -------  --------  [----------]

Total:                                              379293477  100.00%  38792464  [      0.10]
```

martinsumner commented:
Query profile (2M keys):

```
erlang:'--'/2                                         3121553     0.46    179142  [      0.06]

leveled_penciller:keyfolder/6                         1252079     0.70    270316  [      0.22]

erlang:'++'/2                                         1050774     1.54    595734  [      0.57]

perf_SUITE:'-random_queries/5-fun-0-'/3              40020670     3.62   1405230  [      0.04]

leveled_penciller:'-find_nextkeys/6-fun-1-'/1        38770487     3.74   1449232  [      0.04]

leveled_codec:'-accumulate_index/2-fun-1-'/4         40020670     4.07   1577706  [      0.04]

leveled_penciller:compare_nextkey/2                  55594956     7.51   2911541  [      0.05]

maps:update_with/3                                   40021618     7.94   3079187  [      0.08]

leveled_codec:maybe_accumulate/5                     41271801     7.99   3097875  [      0.08]

leveled_penciller:find_nextkeys/6                    98779474    20.73   8040614  [      0.08]

erlang:binary_to_term/1                                786535    36.49  14157118  [     18.00]

--------------------------------------------------  ---------  -------  --------  [----------]

Total:                                              379293477  100.00%  38792464  [      0.10]
```

martinsumner committed:
Standardise use of key_dominates

Use key_dominates by first checking guards, and only on equal keys calling leveled_codec:key_dominates function. This process should be consistent between leveled_penciller:find_nextkeys (used in queries/folds), and leveled_sst:merge_lists (used in ledger merges prompted by load.

martinsumner committed:
Change fullPerf

Change the fullPerf test to run more tests, but with fewer keys.

Given that RS of 512 is being pushed in Riak, 2M objects is till a 300M+ object cluster. 10M >> 1B.  so these are still reasonable sizes to test.

A profilePerf test also added to generate all the profiles base don 2M objects.

ThomasArts commented on src/leveled_codec.erl:
Looking at the `maybe_accumulate` function. The matching is tricky.
You have the following match pattern:
```
maybe_accumulate(
        [{K, {_SQN, {active, TS}, _SH, _MD, undefined}=V}|T],
        Acc, Count, {Now, _ModRange}=Filter, AccFun)
        when TS >= Now ->
    maybe_accumulate(T, AccFun(K, V, Acc), Count + 1, Filter, AccFun);
...
maybe_accumulate(
        [{K, {_SQN, {active, TS}, _SH, _MD, LMD}=V}|T],
        Acc, Count, {Now, {LowDate, HighDate}}=Filter, AccFun)
        when TS >= Now, LMD >= LowDate, LMD =< HighDate ->
    maybe_accumulate(T, AccFun(K, V, Acc), Count + 1, Filter, AccFun);
```

    
But `_ModRange` also matches a tuple, so in case you have a date range but LMD is undefined, then you get the first alternative. Which is no problem, since it results in the same code. But it reads confusing. One wonders... is this undefined a special case and if ModRange is not undefined, is it then fine to tread the ModRange as a Low and High date?

Possibly a comment here in style of: "In case of LMD not undefined, then we are in case of date segment. But if undefined, then must be a ModRange"

ThomasArts commented on src/leveled_penciller.erl:
```diff
@@ -1613,338 +1615,236 @@ compare_to_sqn(ObjSQN, _SQN) when is_integer(ObjSQN) ->
+    :: #{
+        leveled_pmanifest:lsm_level() =>
+            list(leveled_sst:expandable_pointer()|leveled_codec:ledger_kv()),
+        -1 =>
```
Not sure you really want to put an integer here. The nice thing of Erlang maps is that you could use an atom like: `filteredL0`(see line 824)

ThomasArts replied:
but -1 is most likely also the thing you use below when you do integer comparison and then it makes sense.

martinsumner replied:
The other keys in the map are integers >=0 representing the level in the lsm tree - but this is the level above level 0 -> hence level -1

ThomasArts commented on src/leveled_penciller.erl:
```diff
@@ -1613,338 +1615,236 @@ compare_to_sqn(ObjSQN, _SQN) when is_integer(ObjSQN) ->
+    % strip these back before starting the fold 
+    StripIMMFun =
+        fun(MemIter) ->
+            lists:reverse(
```
this looks expensive, reverse twice and dropwhile.
Isn't this the same as a takewhile without reverses?

ThomasArts replied:
No it is not the same. The reverse twice isn't that bad at all. I tried a few other ways and was not successful, so the penalty for list reverse in Erlang is very low. No reason to worry.
For example, a `foldr` would be more expensive.

martinsumner replied:
Agree.  I did some local tests and reached the same conclusion.

martinsumner commented:
Comparison between PR an D3.1, averaging over 3 runs - with each run 3 x 2M object keys, 2 x 5M, 1 x 10M.

<img width="1582" alt="Screenshot 2023-12-10 at 22 07 22" src="https://github.com/martinsumner/leveled/assets/1628897/62540c8c-ed46-47e7-ac49-5910207c4b01">


Hard to explain potential degradation in GET, as the GET path has not been changed.  The performance of the GET load is very dependent on the use of page cache - and so may be related to this, as no obvious difference in the `eprof` profile for this process.

martinsumner commented:
Profile for GET on this PR:

```
leveled_cdb:get/7                                    491988     0.52    313910  [      0.64]

leveled_cdb:calc_crc/2                               500000     0.57    341491  [      0.68]

os:timestamp/0                                      4091148     0.58    349755  [      0.09]

crypto:hash_nif/2                                    499995     0.63    380939  [      0.76]

erlang:spawn_link/3                                  500705     0.68    406909  [      0.81]

leveled_tree:search/3                               1532987     0.68    408730  [      0.27]

leveled_tree:iterator_from/3                        5701976     0.69    411710  [      0.07]

rand:uniform/1                                      2534445     0.69    417023  [      0.16]

gen_server:loop/7                                   1500077     0.74    442837  [      0.30]

prim_buffer:size/1                                  4811555     0.76    457245  [      0.10]

erlang:demonitor/2                                  2542417     0.76    458615  [      0.18]

prim_file:read/2                                    2708252     0.76    458649  [      0.17]

leveled_sst:fetch/12                                 542421     0.77    462934  [      0.85]

gen_statem:loop_receive/3                           1042421     0.92    552166  [      0.53]

gen:reply/2                                         2542416     1.21    728396  [      0.29]

leveled_ebloom:check_hash/2                         1073072     1.30    783075  [      0.73]

leveled_tree:lookup_best/2                          9705617     1.37    822975  [      0.08]

prim_file:pread_nif/3                                500705     2.21   1325558  [      2.65]

gen:do_call/4                                       2542417     2.55   1530388  [      0.60]

prim_file:seek_nif/3                                2103303     2.73   1638351  [      0.78]

leveled_util:hash1/2                               27999946     2.90   1744528  [      0.06]

leveled_sst:find_pos/4                             67052786     4.29   2580094  [      0.04]

zlib:inflate_nif/4                                   500000     8.04   4834770  [      9.67]

prim_file:read_nif/2                                2708252     8.68   5217268  [      1.93]

erlang:binary_to_term/1                             1097642    25.82  15518315  [     14.14]

------------------------------------------------  ---------  -------  --------  [----------]

Total:                                            360232124  100.00%  60102694  [      0.17]
```

Profile on GET for D3.1:

```
erlang:spawn_link/3                                  500986     0.52    324094  [      0.65]

crypto:hash_nif/2                                    499992     0.52    325448  [      0.65]

os:timestamp/0                                      4242883     0.53    326948  [      0.08]

leveled_tree:iterator_from/3                        5765500     0.62    383920  [      0.07]

leveled_tree:search/3                               1606414     0.65    401810  [      0.25]

prim_buffer:size/1                                  5043136     0.65    405213  [      0.08]

rand:uniform/1                                      2609013     0.65    406026  [      0.16]

leveled_sst:fetch/12                                 610896     0.66    411607  [      0.67]

gen_server:loop/7                                   1500074     0.68    425390  [      0.28]

prim_file:read/2                                    2831435     0.72    448924  [      0.16]

erlang:demonitor/2                                  2610889     0.72    449406  [      0.17]

gen_statem:loop_receive/3                           1110896     0.83    514302  [      0.46]

leveled_tree:lookup_best/2                          9726795     1.11    688052  [      0.07]

gen:reply/2                                         2610888     1.12    694290  [      0.27]

leveled_ebloom:check_hash/2                         1259131     1.22    760691  [      0.60]

prim_file:pread_nif/3                                500986     1.72   1067908  [      2.13]

gen:do_call/4                                       2610889     2.22   1384324  [      0.53]

prim_file:seek_nif/3                                2211701     2.31   1436109  [      0.65]

leveled_util:hash1/2                               27999940     2.78   1731324  [      0.06]

leveled_sst:member_check/2                         78135669     4.71   2929871  [      0.04]

prim_file:read_nif/2                                2831435     6.67   4153053  [      1.47]

zlib:inflate_nif/4                                   500000     7.48   4658322  [      9.32]

leveled_sst:find_pos/4                             78747192     9.45   5879562  [      0.07]

erlang:binary_to_term/1                             1118804    24.47  15229522  [     13.61]

------------------------------------------------  ---------  -------  --------  [----------]

Total:                                            455898954  100.00%  62250132  [      0.14]
```

The profile is equivalent, except that with this PR the call to leveled_sst:find_pos/4 is not required.

martinsumner commented:
The perfSUITE load tests shows a possible marginal gain during data load.  The potential for a marginal gain is supported by the fprof profile.

Profile on LOAD for this PR (threshold 0.5% of CPU of more):

```
ets:match_object/2                                           321     0.57    352742  [   1098.88]

dict:fold_seg/4                                          3400000     0.60    373274  [      0.11]

leveled_sst:binaryslot_reader/6                            72435     0.60    373455  [      5.16]

prim_file:sync_nif/2                                         286     0.70    438460  [   1533.08]

leveled_codec:to_lookup/1                               10583879     0.71    440578  [      0.04]

zlib:deflateInit_nif/6                                    200000     0.75    464679  [      2.32]

erlang:term_to_binary/1                                  1036346     0.88    544974  [      0.53]

leveled_codec:count_tombs/2                             13689282     0.94    586147  [      0.04]

leveled_sst:serialise_block/2                             362267     0.97    601781  [      1.66]

leveled_sst:key_dominates/3                             13628396     0.99    616437  [      0.05]

gen:do_call/4                                             608089     1.08    673129  [      1.11]

lists:last/2                                            15961997     1.08    675065  [      0.04]

lists:split/3                                           14174036     1.19    743055  [      0.05]

leveled_util:hash1/2                                    11200000     1.31    813085  [      0.07]

lists:ukeymerge2_1/7                                     8064032     1.40    873485  [      0.11]

leveled_sst:deserialise_checkedblock/2                    340698     1.49    925603  [      2.72]

leveled_ebloom:get_mask/1                                3249980     1.49    926544  [      0.29]

leveled_sst:accumulate_positions/2                       4024779     1.65   1028494  [      0.26]

leveled_ebloom:add_hashlist/34                           2972858     1.66   1030391  [      0.35]

prim_file:pwrite_nif/3                                    200000     1.79   1115543  [      5.58]

lists:append/2                                             72288     1.86   1154930  [     15.98]

erlang:'++'/2                                             594584     2.23   1386651  [      2.33]

leveled_sst:key_dominates_expanded/2                    13624118     2.39   1490231  [      0.11]

ets:insert/2                                              400000     2.52   1567385  [      3.92]

leveled_sst:form_slot/7                                 13689309     2.56   1590582  [      0.12]

zlib:deflate_nif/4                                        200000     8.55   5322068  [     26.61]

erlang:binary_to_term/1                                   540932    11.48   7143701  [     13.21]

erlang:term_to_binary/2                                   726207    28.17  17530440  [     24.14]

-----------------------------------------------------  ---------  -------  --------  [----------]

Total:                                                 215072545  100.00%  62223523  [      0.29]
```

Profile for load on develop-3.1 shows the refcatoring of `leveled_sst:key_dominates_expanded/2` reduces workk split across multiple `key_dominates` and functions (e.g. `maybe_reap_expired_key/2`, `strip to status_only/1`).  Also the process of counting tombs - the `leveled_codec:count_tombs/2` function replaces calls to `lists:foldl_1/3` and `leveled_sst:'-count_tombs/2-fun-0-'/2 `:

```
zlib:deflateInit_nif/6                                    200000     0.63    417043  [      2.09]

leveled_sst:binaryslot_reader/6                            75462     0.66    434472  [      5.76]

leveled_codec:to_lookup/1                               12599384     0.71    465890  [      0.04]

erlang:term_to_binary/1                                  1038189     0.75    495818  [      0.48]

prim_file:sync_nif/2                                         313     0.77    507851  [   1622.53]

leveled_codec:maybe_reap/2                              14984385     0.82    541196  [      0.04]

leveled_sst:serialise_block/2                             377765     0.84    558141  [      1.48]

gen:do_call/4                                             609162     0.89    589472  [      0.97]

lists:last/2                                            17382815     1.00    661578  [      0.04]

lists:ukeymerge2_1/7                                     6944624     1.04    686784  [      0.10]

leveled_ebloom:add_hashlist/34                           2281446     1.04    687130  [      0.30]

leveled_ebloom:get_mask/1                                2593337     1.04    690360  [      0.27]

lists:split/3                                           15253135     1.06    698095  [      0.05]

leveled_util:hash1/2                                    11200000     1.13    745982  [      0.07]

leveled_sst:accumulate_positions/2                       3380205     1.23    810504  [      0.24]

prim_file:pwrite_nif/3                                    200000     1.29    849852  [      4.25]

lists:append/2                                             79756     1.31    867569  [     10.88]

leveled_sst:deserialise_checkedblock/2                    354899     1.32    869670  [      2.45]

leveled_sst:'-count_tombs/2-fun-0-'/2                   14984249     1.68   1113078  [      0.07]

leveled_sst:maybe_expand_pointer/1                      29968771     1.69   1120032  [      0.04]

leveled_codec:maybe_reap_expiredkey/2                   14984385     1.74   1146800  [      0.08]

lists:foldl_1/3                                         16425519     1.83   1207559  [      0.07]

leveled_codec:strip_to_statusonly/1                     29968634     1.85   1220036  [      0.04]

erlang:'++'/2                                            1359388     1.95   1289314  [      0.95]

leveled_codec:key_dominates/2                           13832421     2.21   1460661  [      0.11]

ets:insert/2                                              400000     2.22   1466507  [      3.67]

leveled_sst:form_slot/7                                 15052315     2.47   1633249  [      0.11]

leveled_sst:key_dominates_expanded/3                    14984385     2.48   1640783  [      0.11]

leveled_sst:key_dominates/3                             14984386     2.59   1709923  [      0.11]

zlib:deflate_nif/4                                        200000     7.86   5192417  [     25.96]

erlang:binary_to_term/1                                   555150    10.70   7072659  [     12.74]

erlang:term_to_binary/2                                   739889    25.08  16573265  [     22.40]

-----------------------------------------------------  ---------  -------  --------  [----------]

Total:                                                 342073622  100.00%  66078765  [      0.19]
```

martinsumner commented:
Note that in the LOAD profile, 1.31% is `lists:append/2`.  This is in:

https://github.com/martinsumner/leveled/blob/9f70bfaa1835a4d8c3d896817768eb71f3a56e85/src/leveled_sst.erl#L1731

In this case, the code is doing lists:append(LargeList, SmallList), which is against the guidance - https://www.erlang.org/doc/efficiency_guide/listhandling

i.e.:

```
Eshell V14.1 (press Ctrl+G to abort, type help(). for help)
1> SmallList = lists:seq(1, 256).
[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,
 23,24,25,26,27,28,29|...]
2> LargeList = lists:seq(1, 8096).
[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,
 23,24,25,26,27,28,29|...]
3> timer:tc(fun() -> lists:append(SmallList, LargeList) end).
{1,
 [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,
  23,24,25,26,27|...]}
4> timer:tc(fun() -> lists:append(LargeList, SmallList) end).
{56,
 [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,
  23,24,25,26,27|...]}
```

As the final Hashlist is passed to `leveled_ebloom:create_bloom/1` which does not care about order, this is needlessly inefficient.

ThomasArts commented on src/leveled_sst.erl:
```diff
@@ -1946,45 +1948,51 @@ lookup_slots(StartKey, EndKey, Tree, FilterFun) ->
-                        LMDAcc0};
+                    accumulate_positions(
+                        T,
+                        {<<(32768 + PosH1):16/integer, PosBin/binary>>,
```
This is correct, but I found the previous code more readable. In order to validate correctness, one really needs to know what extract_hash is doing, whereas in the earlier code it was clear that a new integer was created, no matter what 16 bit PosH1 was provided.

martinsumner replied:
I've changed the tune_hash/1 function to return a 2-byte hash between 2 ^ 15 and 2 ^16 - 1, to make this a bit more explicit.

I'd seen some (inconclusive) evidence from profiling that there was an improvement when working in full bytes.  As part of the change to using a 2-byte hash, the find_pos function has been refactored to match instead on whole bytes.

The combination of this and splitting the find_pos function to be different when matching on multiple or single hashes - improved the time taken by find_pos in both Head load and AAE fold load (noticeably in the latter by > 10%).

ThomasArts commented on src/leveled_sst.erl:
```diff
@@ -1946,45 +1948,51 @@ lookup_slots(StartKey, EndKey, Tree, FilterFun) ->
-                        LMDAcc0}
+                    accumulate_positions(
+                        T,
+                        {<<(32768 + PosH1):16/integer,
```
same as above

ThomasArts commented on src/leveled_sst.erl:
```diff
@@ -1946,45 +1948,51 @@ lookup_slots(StartKey, EndKey, Tree, FilterFun) ->
+                    accumulate_positions(
+                        T,
+                        {<<(32768 + PosH1):16/integer,
+                                NHC:8/integer,
```
NHC:8 is clearer than previous code, because there is a comment saying that NHC is at most 127

ThomasArts reviewed (approved):
Considerable performance improvements and code gets better as well.

Ran the Quickcheck tests just to be sure as well as dialyzer and test.
All good.

martinsumner committed:
Mas d31 nhskv16sstpclv2 (#427)

* Change fullPerf

Change the fullPerf test to run more tests, but with fewer keys.

Given that RS of 512 is being pushed in Riak, 2M objects is till a 300M+ object cluster. 10M >> 1B.  so these are still reasonable sizes to test.

A profilePerf test also added to generate all the profiles base don 2M objects.

* Address BigList ++ SmallList

Within the SST path there was use of ++ or lists:append where potentially lists:append(BigList, SmallList) was being used when lists:append(SmallList, BigList) would be more efficient.

The fetching of trimmed blocks in queries is now change to not use a large fold function, and now uses blocks_required/7 - where further efficiency is made by guarding against deserialisation of blocks 1 and 5 if they are out of the range.

Most of the performance improvement is in the building of the hashlist.

The other refactoring causes some confusion due to the returning of revrsed lists abd binaryslot_trimmedlist has been renamed to binaryslot_trimmedreversed to reflect this.

There is a functional change in that calls to binaryslot_trimmedlist with a range and a SegCheckFun would not previously be trimmed - whereas they should be.

* Alter ++ order in binaryslot_reader

Normally SlotsTOPoint will be smaller, so prepend this to the fetched list.

This will allow for larger ITERATOR_SCANWIDTH at lower cost, potentially.

* Reduce duplication

Duplication between sst_getfilteredslots and sst_getfilteredrange.

Tries hard to avoid appending a short list to a long list.

Also ensures that through all paths results are trimmed to both the left and right.

* Extend test

Queries where previously all returning a large number of index entries - changes made to make number of entries per query more realistic.  Also an update process added to show difference between loading and rotating keys.

* Trim before accumulating

When deserialising blocks, trim them at that point prior to accumulation.

Standardise the functions used for trimming, and add a unit test to ensure that the accumulation handles empty blocks correctly.

* Relabel as AAE fold

* Review/update comments

* Reset default test

* Remove redundant define statements

* Increase timeout

The clerk may be doing a large fold - and 20s may not be sufficient.  Has been seen to fail to shut down at the end of perf_SUITE fullperf test.

* Test v5

Test mini-queries - where generally a small number of entries are returned

* Remove unnecessary case

Single statement works for all N > 1

* Revert to ctperf as default

Bump timeout for slower AWS instances

martinsumner committed:
Tune hash to be between 2^15 and 2 ^16

Allows for find_pos to match directly on 2-byte integer.

The profile indicates that this improves the speed of find_pos, as does having a separate function for finding an individual hash or multiple hashes.

martinsumner commented:
This chart is a comparison of running the perf_SUITE riak_fullperf test comparing the develop-3.1 branch with LZ4 compression against various alternative setups (including this PR).  This PR is referred here to as PR427 (which has now been merged into PR426)

![PR426 Testing](https://github.com/martinsumner/leveled/assets/1628897/7b2369ab-93c2-415e-bf6a-1b886d0ed671)

The 100% line relates to performance with LZ4 compression and develop-3.1 on OTP 22.

martinsumner committed:
DJB Hash and integer overflow

The implementation of magic_hash was a straight interpretation of the DJB hash code.  Every byte of the key in turn is bxor'd against an accumulated hash (which is multiplied by 33 (i.e. bsl 5 plus itself) every loop.

In the original C code, due to the integer overflow of unsigned integers, every loop the extra bits beyond 4 bytes in the hash are automatically discarded.  However, in Erlang the integer type can be of an arbitrary size - so each loop carries a larger hash.  This means that with a 100-byte key (say), there would be a 500-bit integer hash by the end (which is then chopped back to 32 bits before returning).

This change explicitly restricts the hast to a 32-bit integer on every loop.  This doesn't change the result, but is more efficient (and this can be detected in the perf_SUITE profile test, especially when using.a larger bucket name to generate larger keys).

martinsumner committed:
Avoid double-reverse and lists:last

The generate_binary_slot function is primarily called from within form_slot/7.  In this case form_slot/7 would have the slot (i.e. the list of keys to form the slot) in reverse, and so would reverse it before calling generate_binary_slot.  If the slot was a lookup type, the generate_binary_slot function would then re-reverse to be efficient on accumulate_positions.  It would also need the last key - which it would find through lists:last.

Now the generate_binary_slot function can receive the slot reversed, and so will only reverse once.  As it will always reverse once, it will always have the list in reverse and so can find the last key through the more efficient hd/1 function.

When forming level 0 there will now be an additional lists:reverse/1 call - but normally there are an order of magnitude more non-level-0 file creations for each level-0 creation ... so this is still a net gain.

The BuildTimings was originally used to try and profile this in production - which seems unnecessary given the ability to profile via perf_SUITE. This is also altered to reduce the situations where this is called, and reduce the number of required calls to os:timestamp/1 to make it.

martinsumner committed:
Make key_dominates more efficient

the most common case is when the created file is not. abasement file and the key in list 2 is < key in list 1.  so avoid the separate function call to key_dominates_expanded in this situation.

The load profile prior to this change has:

leveled_sst:key_dominates/3                             16139172     1.60    820906  [      0.05]

leveled_sst:key_dominates_expanded/2                    16133866     3.62   1859403  [      0.12]

So for every key covered by every merge, there is a 0.05 microsecond cost on key_dominates/3, and and a 0.12 microsecond cost on key_dominates_expanded/2.

Following the change, the profile is:

leveled_sst:key_dominates_expanded/2                     1621644     0.20    103125  [      0.06]

leveled_sst:key_dominates/3                             16235173     3.22   1663231  [      0.10]

Now key_dominates_expanded/2 is called an order of magnitude less time, and key_dominates/3 is more efficient than the previous combined cost.

martinsumner committed:
leveled_ebloom efficiency

Improve leveled_ebloom efficiency, by no longer handling super-large (i.e. 512 byte) integers.

This is a significant speed-up when checking, and a potential speed-up when building (certainly as measured by the leveled_ebloom unit test).

Prior to change:

```
Test with size 80000 has microsecond timings: - build 27982.0 check 49384.25 neg_check 48557.75 and fpr 0.20890195845586051
Test with size 60000 has microsecond timings: - build 18112.25 check 34294.25 neg_check 34416.0 and fpr 0.13369656503699867
Test with size 40000 has microsecond timings: - build 12193.0 check 19180.25 neg_check 18963.5 and fpr 0.06915950943254698
Test with size 32768 has microsecond timings: - build 9858.5 check 16857.0 neg_check 15792.75 and fpr 0.10147250784345158
Test with size 20000 has microsecond timings: - build 4317.5 check 9586.0 neg_check 9575.75 and fpr 0.068836686242659
Test with size 10000 has microsecond timings: - build 1640.5 check 4676.0 neg_check 4609.0 and fpr 0.06851578341302225
Test with size 5000 has microsecond timings: - build 1112.5 check 2396.75 neg_check 2392.25 and fpr 0.01979951124632188
Test with size 2000 has microsecond timings: - build 402.75 check 952.0 neg_check 931.75 and fpr 0.121131741821397
```

Post change:

```
Test with size 80000 has microsecond timings: - build 16301.5 check 3165.25 neg_check 3417.75 and fpr 0.2078829130796741
Test with size 60000 has microsecond timings: - build 13206.25 check 2406.0 neg_check 2535.5 and fpr 0.13489507929376135
Test with size 40000 has microsecond timings: - build 7730.25 check 1298.5 neg_check 1307.25 and fpr 0.06991004956962564
Test with size 32768 has microsecond timings: - build 7006.0 check 1185.25 neg_check 1224.0 and fpr 0.1279901844979081
Test with size 20000 has microsecond timings: - build 4258.5 check 814.5 neg_check 720.0 and fpr 0.0684979441868603
Test with size 10000 has microsecond timings: - build 1469.25 check 321.75 neg_check 316.0 and fpr 0.0692938177819981
Test with size 5000 has microsecond timings: - build 729.25 check 177.75 neg_check 178.75 and fpr 0.06775364119898593
Test with size 2000 has microsecond timings: - build 290.5 check 73.25 neg_check 74.25 and fpr 0.14997483643683945
```

martinsumner committed:
Make bloom fixed FPR

The previous implementation restricted sizes to factors of 2 (then multiplied by 512 bytes).  There was no reason to be limited to factors of 2 though.

This implementation is backwards compatible, it will read a bloom filter still of any size - but will right-size the filter close to the target fpr of 5% when building a new filter.

Comments now added, with further clarification, and the log output changed to make it clear what the timings relate to.

On a 2021 M1 - check times are now 4 hundredths of a microsecond:

Test with size 250000 has microsecond timings: - build in 59545 then 0.034 per pos-check, 0.039 per neg-check, fpr 0.379 with bits-per-key 0.131
Test with size 80000 has microsecond timings: - build in 20259 then 0.036 per pos-check, 0.036 per neg-check, fpr 0.069 with bits-per-key 0.819
Test with size 60000 has microsecond timings: - build in 11874 then 0.033 per pos-check, 0.034 per neg-check, fpr 0.053 with bits-per-key 0.998
Test with size 40000 has microsecond timings: - build in 7869 then 0.043 per pos-check, 0.034 per neg-check, fpr 0.051 with bits-per-key 0.998
Test with size 32768 has microsecond timings: - build in 6741 then 0.044 per pos-check, 0.039 per neg-check, fpr 0.049 with bits-per-key 1.000
Test with size 20000 has microsecond timings: - build in 5399 then 0.037 per pos-check, 0.032 per neg-check, fpr 0.051 with bits-per-key 0.992
Test with size 10000 has microsecond timings: - build in 1591 then 0.032 per pos-check, 0.032 per neg-check, fpr 0.049 with bits-per-key 0.973
Test with size 5000 has microsecond timings: - build in 748 then 0.036 per pos-check, 0.036 per neg-check, fpr 0.058 with bits-per-key 0.922
Test with size 2000 has microsecond timings: - build in 277 then 0.032 per pos-check, 0.032 per neg-check, fpr 0.071 with bits-per-key 0.768
Test with size 1000 has microsecond timings: - build in 137 then 0.035 per pos-check, 0.032 per neg-check, fpr 0.049 with bits-per-key 1.024

martinsumner committed:
Correct to bytes not bits

ThomasArts reviewed (approved):
wrote a simple quickcheck property that confirms that the result is the same.

ThomasArts commented on src/leveled_sst.erl:
```diff
@@ -2735,17 +2742,22 @@ split_lists(KVList1, SlotLists, N, PressMethod, IdxModDate) ->
-                InitTombCount,
-                #build_timings{}).
+    BuildTimings = 
+        case IsBase or lists:member(L, ?LOG_BUILDTIMINGS_LEVELS) of
```
if we talk efficiency:  use `orelse` instead of `or` here. No need to check membership if `IsBase` is `true`.
On the other hand checking member of singleton list is rather cheap.

martinsumner replied:
Fixed

martinsumner commented:
Relative performance for this PR to develop-3.1.  Comparing both OTP22, and further improvements with OTP 26.

<img width="1749" alt="Relative Performance PR426 OTP22:26" src="https://github.com/martinsumner/leveled/assets/1628897/73889663-ee3f-429c-a1ae-d69ef73d2468">

ThomasArts commented on src/leveled_sst.erl:
```diff
@@ -2950,37 +2960,44 @@ key_dominates_expanded([H1|T1], [H2|T2]) ->
-    Timer = timer:now_diff(os:timestamp(), SW),
+update_buildtimings(Timings, Stage) ->
+    LastTS = Timings#build_timings.last_timestamp,
+    ThisTS = os:timestamp(),
```
up to taste, but would it not be nicer to say here: `NewTimings = Timings#build_timings{last_timestamp = os:timestamp()}` and then in all cases below use `NewTimings#build_timings{...}` where you only update the thing that matters, not the timestamp.

martinsumner replied:
Fixed

ThomasArts commented on src/leveled_ebloom.erl:
```diff
@@ -4,20 +4,24 @@
+-define(MASK_BAND, 63).
+    % i.e. integer slize size - 1
+-define(SPLIT_BAND, 4095).
+    % i.e. (?BLOOM_SLOTSIZE_BYTES div 8) - 1
```
`% i.e. (?BLOOM_SLOTSIZE_BYTES div 8) - 1`
If I compute correctly: `% i.e. (?BLOOM_SLOTSIZE_BYTES * 8) - 1`

martinsumner replied:
Fixed

ThomasArts commented on src/leveled_ebloom.erl:
```diff
@@ -31,62 +35,45 @@
+                16;
+            L when L > 4096 ->
+                8;
+            L when L > 4096 ->
```
L > 2048 ??

ThomasArts commented on src/leveled_ebloom.erl:
```diff
@@ -31,62 +35,45 @@
+        map_hashes(
+            HashList,
+            list_to_tuple(
+                lists:map(fun(_) -> [] end, lists:seq(1, SlotCount))),
```
`lists:duplicate(SlotCount, [])`

martinsumner replied:
Fixed

ThomasArts reviewed (changes requested):
Line 50 states L > 4096 (but should be 2048).
However, this is not found in your tests... it means that whenever your Hashlist has a length betwen 2048 and 4096, you make the wrong SlotHashes... but you seem not to detect that while testing.

Would be good to understand why this can slip tests

ThomasArts commented on src/leveled_pmem.erl:
```diff
@@ -71,7 +71,7 @@ prepare_for_index(IndexArray, no_lookup) ->
     {Slot, H0} = split_hash(Hash),
     Bin = array:get(Slot, IndexArray),
-    array:set(Slot, <<Bin/binary, 1:1/integer, H0:23/integer>>, IndexArray).
+    array:set(Slot, <<Bin/binary, H0:24/integer>>, IndexArray).
```
This change suggests that split_hash always returns a `H0` in which the most significant bit is set. I believe you but hard to digest from the code.

martinsumner replied:
It is just that previously we were checking for the leading bit of 1 (in find_pos/2), but it had no meaning in this function (i.e. there was no case for the leading bit of 0).

So we were needlessly matching on just 23 bits of the hash, when we could be doing a simpler match on 24 bits.  The hash can have a leading bit of 0, as find_pos/2 doesn't care.

ThomasArts commented on src/leveled_pmem.erl:
```diff
@@ -201,16 +201,16 @@ merge_trees(StartKey, EndKey, TreeList, LevelMinus1) ->
 find_pos(<<>>, _Hash) ->
     false;
-find_pos(<<1:1/integer, Hash:23/integer, _T/binary>>, Hash) ->
+find_pos(<<Hash:24/integer, _T/binary>>, Hash) ->
```
This can now return `true` if the hash's most significant bit it `0`. In the original code, it would then fail. But in the original code, the bit is set to 1 in case of prepare_for_index... a bit hard to grasp what is going on here

martinsumner replied:
The prepending of "1" had no value ... probably a legacy of the related to the use of find_pos in `leveled_sst`.

ThomasArts commented on src/leveled_ebloom.erl:
```diff
@@ -33,24 +42,17 @@
-            _L ->
-                2
+            L ->
+                min(128, max(2, (L - 1) div 512))
```
Solves the problem of missing 2048

martinsumner committed:
Update following review

martinsumner commented:
Updated profile (with addition of ZSTD compression).

LOAD:

```
gen_server:loop/7                                         405933     0.53    242471  [      0.60]

dict:fold_seg/4                                          3400000     0.55    250895  [      0.07]

leveled_codec:maybe_reap_expiredkey/2                    3686698     0.61    276035  [      0.07]

leveled_sst:key_dominates_expanded/2                     3716302     0.64    289182  [      0.08]

lists:reverse/2                                          1158365     0.68    310092  [      0.27]

leveled_sst:update_buildtimings/2                         298044     0.86    388781  [      1.30]

ets:match_object/2                                           322     1.01    456729  [   1418.41]

erlang:'++'/2                                            2639533     1.03    465402  [      0.18]

leveled_codec:to_lookup/1                               13998843     1.10    498513  [      0.04]

prim_file:sync_nif/2                                         298     1.11    503865  [   1690.82]

leveled_codec:count_tombs/2                             15840468     1.30    587822  [      0.04]

gen:do_call/4                                             608373     1.39    632941  [      1.04]

leveled_sst:accumulate_positions/2                       2816117     1.41    637945  [      0.23]

leveled_ebloom:map_hashes/3                              1978514     1.49    676882  [      0.34]

lists:split/3                                           15890381     1.50    681817  [      0.04]

leveled_util:hash1/2                                    20400000     1.77    802293  [      0.04]

lists:ukeymerge2_1/7                                     7668499     2.10    952035  [      0.12]

prim_file:pwrite_nif/3                                    200000     2.25   1020964  [      5.10]

leveled_sst:serialise_block/2                             382304     2.50   1134513  [      2.97]

zstd:quick_decompress/1                                   320722     2.85   1292064  [      4.03]

ets:insert/2                                              400000     2.94   1332114  [      3.33]

leveled_sst:deserialise_checkedblock/2                    359809     3.23   1466165  [      4.07]

leveled_sst:key_dominates/3                             15806346     3.46   1571526  [      0.10]

erlang:term_to_binary/2                                   400233     3.49   1582410  [      3.95]

leveled_sst:form_slot/7                                 15870201     3.69   1675805  [      0.11]

erlang:term_to_binary/1                                  1382369     8.83   4007575  [      2.90]

erlang:binary_to_term/1                                   560041     9.37   4251710  [      7.59]

zstd:quick_compress/2                                     542975    14.73   6683332  [     12.31]

-----------------------------------------------------  ---------  -------  --------  [----------]

Total:                                                 236308443  100.00%  45373633  [      0.19]
```

HEAD:

```
leveled_monitor:maybe_time/1                             3087458     0.52    333024  [      0.11]

dict:find_val/2                                          8911321     0.54    343281  [      0.04]

leveled_sst:spawn_check_block/3                          1001376     0.58    367323  [      0.37]

rand:exsss_uniform/2                                     5087535     0.59    378468  [      0.07]

leveled_pmanifest:key_lookup_level/3                     6030762     0.62    397271  [      0.07]

os:timestamp/0                                           5875556     0.63    401170  [      0.07]

ets:lookup/2                                             1000000     0.64    411399  [      0.41]

leveled_penciller:fetch/5                                4030762     0.71    453167  [      0.11]

gen_statem:loop_receive/3                                1087796     0.76    487655  [      0.45]

leveled_bookie:handle_call/3                             1000000     0.78    495575  [      0.50]

rand:seed_put/1                                          5087562     0.78    500492  [      0.10]

leveled_ebloom:match_hash/3                              3680291     0.79    507345  [      0.14]

erlang:demonitor/2                                       3097831     0.87    552394  [      0.18]

crypto:hash_nif/2                                        1000000     0.87    554133  [      0.55]

gen_server:loop/7                                        2010259     0.88    562030  [      0.28]

erlang:spawn_link/3                                      1001376     0.88    563469  [      0.56]

prim_file:read_nif/2                                       18270     0.97    620767  [     33.98]

lists:foldl_1/3                                          9002775     0.98    628448  [      0.07]

leveled_tree:search/3                                    3108128     1.16    739960  [      0.24]

array:get/2                                             15111145     1.16    740990  [      0.05]

leveled_sst:fetch/12                                     1077366     1.16    743311  [      0.69]

rand:uniform/1                                           5087535     1.18    751065  [      0.15]

gen:reply/2                                              3097744     1.34    854752  [      0.28]

leveled_tree:iterator_from/3                            11690675     1.36    866877  [      0.07]

leveled_pmem:'-check_index/2-fun-0-'/4                   9000000     1.64   1046258  [      0.12]

array:get_1/3                                           36265894     2.22   1419625  [      0.04]

leveled_tree:lookup_best/2                              20484142     2.31   1472974  [      0.07]

leveled_pmem:find_pos/2                                 30706032     2.95   1885049  [      0.06]

prim_file:pread_nif/3                                    1001542     3.01   1922966  [      1.92]

gen:do_call/4                                            3097824     3.58   2288412  [      0.74]

leveled_sst:find_posint/4                              138977166     7.97   5090613  [      0.04]

zstd:quick_decompress/1                                  1013478    13.14   8387379  [      8.28]

erlang:binary_to_term/1                                  1013485    17.60  11233965  [     11.08]

-----------------------------------------------------  ---------  -------  --------  [----------]

Total:                                                 575765126  100.00%  63839782  [      0.11]
```

GET:

```
array:get/2                                         7554667     0.50    335278  [      0.04]

leveled_tree:search/3                               1554277     0.55    374074  [      0.24]

rand:uniform/1                                      2538775     0.57    383108  [      0.15]

leveled_sst:fetch/12                                 538734     0.58    393427  [      0.73]

prim_file:read/2                                    2707195     0.62    420376  [      0.16]

gen_server:loop/7                                   1500082     0.63    427933  [      0.29]

leveled_tree:iterator_from/3                        5845194     0.66    443721  [      0.08]

erlang:demonitor/2                                  2538735     0.67    449567  [      0.18]

gen_statem:loop_receive/3                           1038785     0.68    459338  [      0.44]

leveled_pmem:'-check_index/2-fun-0-'/4              4500000     0.77    516622  [      0.11]

gen:reply/2                                         2538734     0.98    660592  [      0.26]

array:get_1/3                                      18129685     1.05    710876  [      0.04]

leveled_tree:lookup_best/2                         10248708     1.17    791114  [      0.08]

prim_file:pread_nif/3                                500747     1.41    952309  [      1.90]

leveled_pmem:find_pos/2                            15357427     1.56   1053115  [      0.07]

prim_file:seek_nif/3                                2107161     2.33   1569601  [      0.74]

gen:do_call/4                                       2538735     2.45   1656407  [      0.65]

leveled_util:hash1/2                               50999940     2.78   1873591  [      0.04]

leveled_sst:find_posint/4                          69494800     3.79   2561012  [      0.04]

erlang:binary_to_term/1                             1100781     8.42   5680438  [      5.16]

zstd:quick_decompress/1                             1000747     9.90   6678720  [      6.67]

prim_file:read_nif/2                                2707195    34.18  23066983  [      8.52]

------------------------------------------------  ---------  -------  --------  [----------]

Total:                                            414486287  100.00%  67494526  [      0.16]
```

MINi_QUERY:

```
leveled_tree:idxtlookup_range_end/5                    249950     0.50     60876  [      0.24]

lists:map/2                                            851500     0.61     74514  [      0.09]

leveled_sst:'-in_range/3-fun-0-'/2                     997377     0.64     77220  [      0.08]

gen_statem:loop_receive/3                              200083     0.64     77788  [      0.39]

lists:foldl/3                                         1148620     0.79     95835  [      0.08]

leveled_tree:iterator_from/3                           997537     0.81     98578  [      0.10]

lists:reverse/1                                       2400257     0.88    106985  [      0.04]

leveled_penciller:find_nextkeys/6                     1400016     0.89    107927  [      0.08]

erlang:demonitor/2                                     499709     0.99    120254  [      0.24]

lists:takewhile_1/2                                   1447013     1.15    140088  [      0.10]

leveled_tree:'-idxtlookup_range_start/4-fun-0-'/2     1511716     1.20    145337  [      0.10]

lists:splitwith_1/3                                   2008201     1.46    177221  [      0.09]

leveled_tree:skpllookup_to_range/4                     499400     1.51    183524  [      0.37]

gen:reply/2                                            449768     1.61    195047  [      0.43]

leveled_tree:'-skpllookup_to_range/4-fun-1-'/2        2515274     1.63    197874  [      0.08]

prim_file:pread_nif/3                                  100034     1.95    236732  [      2.37]

gen:do_call/4                                          449769     2.31    280170  [      0.62]

leveled_sst:'-in_range/3-fun-1-'/2                    3506002     2.49    302449  [      0.09]

lists:last/2                                          9066288     2.78    337951  [      0.04]

lists:dropwhile_1/2                                   6076818     3.69    448841  [      0.07]

leveled_codec:endkey_passed/2                         3470245     3.91    475153  [      0.14]

zstd:quick_decompress/1                                271037     6.41    778433  [      2.87]

leveled_tree:'-skpllookup_to_range/4-fun-0-'/4       12984400     6.58    799524  [      0.06]

lists:foldl_1/3                                      13933260     7.67    931872  [      0.07]

erlang:binary_to_term/1                                271037    23.41   2844428  [     10.49]

--------------------------------------------------  ---------  -------  --------  [----------]

Total:                                              108409811  100.00%  12148204  [      0.11]
```


FULL-FOLD:

```
gen:do_call/4                                           44321     0.89    590142  [     13.32]

perf_SUITE:'-counter/2-fun-0-'/4                     22400000     1.12    741946  [      0.03]

leveled_head:build_head/2                            22400000     1.19    792202  [      0.04]

leveled_codec:striphead_to_v1details/1               22400000     1.25    829494  [      0.04]

leveled_codec:from_ledgerkey/1                       22400000     1.26    839408  [      0.04]

leveled_penciller:'-find_nextkeys/6-fun-1-'/1        21700000     1.30    863427  [      0.04]

leveled_head:get_size/2                              22400000     1.49    986206  [      0.04]

leveled_codec:to_ledgerkey/3                         22400020     1.59   1056121  [      0.05]

leveled_codec:maybe_accumulate/5                     23100010     2.87   1907389  [      0.08]

maps:update_with/3                                   22400010     3.11   2061331  [      0.09]

leveled_sst:deserialise_checkedblock/2                 872790     5.69   3775442  [      4.33]

leveled_runner:'-accumulate_objects/4-fun-0-'/7      22400000     6.02   3994217  [      0.18]

leveled_head:riak_metadata_to_binary/2               22400000     6.30   4181443  [      0.19]

leveled_codec:return_proxy/4                         22400000     7.13   4732674  [      0.21]

zstd:quick_decompress/1                                872790    12.19   8088984  [      9.27]

erlang:binary_to_term/1                                872790    12.64   8385188  [      9.61]

erlang:term_to_binary/1                              22400000    12.93   8578424  [      0.38]

leveled_penciller:find_nextkeys/6                   127253180    17.38  11532297  [      0.09]

--------------------------------------------------  ---------  -------  --------  [----------]

Total:                                              438196507  100.00%  66361538  [      0.15]
```


GUESS:

```
erlang:spawn_link/3                                    21610     0.57    11689  [      0.54]

leveled_sst:check_blocks/8                            196250     0.57    11762  [      0.06]

lists:sublist_2/2                                     177460     0.62    12825  [      0.07]

lists:last/2                                          368580     0.63    12950  [      0.04]

gen_statem:loop_receive/3                              44292     0.66    13650  [      0.31]

lists:partition_1/4                                   174860     0.69    14177  [      0.08]

lists:foldr_1/3                                       218389     0.72    14779  [      0.07]

leveled_sst:expand_list_by_pointer/5                   44280     0.86    17831  [      0.40]

array:get_1/3                                         519916     0.91    18711  [      0.04]

leveled_sst:checkblocks_segandrange/7                 174640     0.96    19849  [      0.11]

leveled_penciller:find_nextkeys/6                     167443     1.00    20535  [      0.12]

lists:reverse/1                                       559855     1.27    26246  [      0.05]

leveled_sst:extract_header/2                          174640     1.33    27439  [      0.16]

lists:map_1/2                                         349580     1.77    36551  [      0.10]

leveled_sst:'-read_slots/5-fun-0-'/8                  174640     1.91    39289  [      0.22]

erlang:'++'/2                                         295002     2.12    43663  [      0.15]

prim_file:pread_nif/3                                  21610     2.25    46377  [      2.15]

lists:reverse/2                                       178150     2.25    46422  [      0.26]

gen:reply/2                                            44320     2.96    61049  [      1.38]

gen:do_call/4                                          44321     3.20    65934  [      1.49]

zstd:quick_decompress/1                                21610     8.63   177890  [      8.23]

erlang:binary_to_term/1                                21610    11.81   243414  [     11.26]

leveled_sst:find_posmlt/6                           22516070    40.47   834474  [      0.04]

--------------------------------------------------  --------  -------  -------  [----------]

Total:                                              30702338  100.00%  2061897  [      0.07]
```
